go/libraries/doltcore/env/memory.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package env
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/dolthub/dolt/go/store/chunks"
+	"github.com/dolthub/dolt/go/store/hash"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdocs"
+	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
+)
+
+func NewMemoryDbData(ctx context.Context) (DbData, error) {
+	ddb, err := NewMemoryDoltDB(ctx)
+	if err != nil {
+		return DbData{}, err
+	}
+
+	rs, err := NewMemoryRepoState(ctx, ddb)
+	if err != nil {
+		return DbData{}, err
+	}
+
+	return DbData{
+		Ddb: ddb,
+		Rsw: rs,
+		Rsr: rs,
+		Drw: rs,
+	}, nil
+}
+
+func NewMemoryDoltDB(ctx context.Context) (*doltdb.DoltDB, error) {
+	ts := &chunks.TestStorage{}
+	cs := ts.NewViewWithDefaultFormat()
+	ddb := doltdb.DoltDBFromCS(cs)
+
+	m := "memory"
+	b := ref.NewBranchRef("master")
+	err := ddb.WriteEmptyRepoWithCommitTimeAndDefaultBranch(ctx, m, m, doltdb.CommitNowFunc(), b)
+	if err != nil {
+		return nil, err
+	}
+
+	return ddb, nil
+}
+
+func NewMemoryRepoState(ctx context.Context, ddb *doltdb.DoltDB) (MemoryRepoState, error) {
+	branches, err := ddb.GetBranches(ctx)
+	if err != nil {
+		return MemoryRepoState{}, err
+	}
+	head := branches[0]
+
+	rs := MemoryRepoState{
+		DoltDB: ddb,
+		Head:   head,
+	}
+
+	commit, err := ddb.ResolveCommitRef(ctx, head)
+	if err != nil {
+		return MemoryRepoState{}, err
+	}
+
+	root, err := commit.GetRootValue()
+	if err != nil {
+		return MemoryRepoState{}, err
+	}
+
+	err = rs.UpdateWorkingRoot(ctx, root)
+	if err != nil {
+		return MemoryRepoState{}, err
+	}
+
+	err = rs.UpdateStagedRoot(ctx, root)
+	if err != nil {
+		return MemoryRepoState{}, err
+	}
+
+	return rs, nil
+}
+
+type MemoryRepoState struct {
+	DoltDB *doltdb.DoltDB
+	Head   ref.DoltRef
+}
+
+var _ RepoStateReader = MemoryRepoState{}
+var _ RepoStateWriter = MemoryRepoState{}
+var _ DocsReadWriter = MemoryRepoState{}
+
+func (m MemoryRepoState) CWBHeadRef() ref.DoltRef {
+	return m.Head
+}
+
+func (m MemoryRepoState) CWBHeadSpec() *doltdb.CommitSpec {
+	spec, err := doltdb.NewCommitSpec(m.CWBHeadRef().GetPath())
+	if err != nil {
+		panic(err)
+	}
+	return spec
+}
+
+func (m MemoryRepoState) UpdateStagedRoot(ctx context.Context, newRoot *doltdb.RootValue) error {
```

Comment:
Mind adding a TODO to this and the other root update method? I'm in the middle of killing them in the interface, want to make sure to get them here too